### DISCUSSION TITLE
I want to add unified interceptor，wrapper action report error

### DISCUSSION THREAD

**Discussion by hj2060** [remix-run/remix#4134] [Q&A] [answered]
```ts
const sessionAction = (fn: (event: DataFunctionArgs, userId: number) => any): ActionFunction => {
  return (e) => {
    return fn(e, 1)
  }
}
// in route

export const action = sessionAction(async (e) => {
  const form = e.request.formData()
  return json({}, 400)
})
```

browser get error 
<img width="798" alt="image" src="https://user-images.githubusercontent.com/9586285/188423826-3f564e50-7dfa-46db-a540-94ad21ba64ec.png">

**Comment by sergiodxa** [comment, marked as answer]
The problem is that when you wrap your action (or loader) in a high order function, you're potentially creating a side-effect, so the compiler doesn't know if sessionAction does something that the client-side bundle needs. Because of this, the compiler decides to keep the action on the client-side bundle, and because it keeps the action, it also keeps anything else you imported to use there.

This is explained on the [Module Constraints guide](https://remix.run/docs/en/v1/guides/constraints#higher-order-functions) of the documentation.

The solution is that, instead of using a HOF to wrap the action, call a function inside the action, you can throw errors or responses from inside it, so if you want to, for example, redirect the user if condition don't pass, throw a redirect response from your function you call inside the action.

For example, let's say you have a requireUser HOF that you use like this:

```ts
export let action: requireUser(async ({ user }) => {
  // do something with user here
})
```

Here, requireUser will redirect the user to `/login` if it's not authenticated, and it will add `user` to the arguments the action receives if it is authenticated.

Because of the module constraints you could change it to work this way:

```ts
async function requireUser(request: Request) {
  let session = await getSession(request.headers.get("cookie"))
  let userId = session.get("userId")
  if (!userId) throw redirect("/login")
  return await getUserById(userId)
}
```

And you can use it this way

```ts
export async function action({ request }: ActionArgs) {
  let user = await requireUser(request)
  // do something with user here
}
```

This achieve the same thing, if the user is authenticated you get it as the return value, if not it will throw a redirect response to the `/login` route and the rest of your action code will never run.

**Reply by hj2060** [reply]
ok thanks for the detailed answer 👍
